mindmeld/core.py

```diff
@@ -699,25 +700,89 @@ def from_query(
+                span_out (Span): Span of the token of interest in the full_norm_text.
+            Return:
+                tok_start (int): The starting token index.
+            """
             # The span range is till the span_out or max to the second last char
             tok_start = 0
-            span_range = min(span_out.start, len(full_text) - 1)
-            for idx, current_char in enumerate(full_text[:span_range]):
+            span_range = min(span_out.start, len(full_norm_text) - 1)
+            for idx, current_char in enumerate(full_norm_text[:span_range]):
                 # Increment the counter only if a whitespace follows a non-whitespace
-                next_char = full_text[idx + 1]
+                next_char = full_norm_text[idx + 1]
                 if not current_char.isspace() and next_char.isspace():
                     tok_start += 1
-            tok_span = Span(tok_start, tok_start - 1 + len(text.split()))
+            return tok_start
+
+        def _get_token_span(query_span, form_in, offset):
+            """ Get the token span of text based on the normalized token sentence.
+            Token span should be calculated based on the normalized token sentence
+            as this approach is more robust to sentences that are not space delimited.
+
+            Args:
+                query_span (Span): The text span for the token of interest.
+                form_in (int): Integer value representing starting text format.
+                offset (int): Offset the final indexing to parent's indexing.
+            Returns:
+                tok_span (Span): Span of normalized tokens for the given text.
+            """
+            # Creating a copy to avoid modifying the original query object
+            query_for_norm_transformation = deepcopy(query)
+
+            span_out = query_for_norm_transformation.transform_span(
+                query_span, form_in, TEXT_FORM_NORMALIZED
+            )
+            full_norm_text = query_for_norm_transformation.get_text_form(TEXT_FORM_NORMALIZED)
+            norm_token_text = span_out.slice(full_norm_text)
+            tok_start = _get_token_start(full_norm_text, span_out)
+
+            # Using a min token len of 1 avoids token span of (x, x - 1) which can become negative.
+            norm_token_text_len = len(norm_token_text.split()) or 1
+            tok_span = Span(
+                start=tok_start,
+                end=tok_start + norm_token_text_len - 1
+            )
+
+            # convert span from query's indexing to parent's indexing
+            if offset is not None:
+                offset_out = query_for_norm_transformation.transform_index(
+                    offset, form_in, TEXT_FORM_NORMALIZED
+                )
+                tok_offset = len(full_norm_text[:offset_out].split())
+                tok_span.shift(-tok_offset)
+
+            return tok_span
+
+        def _get_form_details(query_span, offset, form_in, form_out):
```

Comment:
I think _get_form_details and _get_token_span can be combined since the `if offset is not None:` cases seem close to each other. There are other ops that are similar as well